apps/aecore/src/aec_dev_reward.erl

```diff
@@ -0,0 +1,122 @@
+            end;
+        {error, Reason} ->
+            lager:error("invalid protocol beneficiaries: ~p", [Reason]),
+            exit({invalid_protocol_beneficiaries, Reason})
+    end.
+
+cfg(Key, Default) ->
+    aeu_env:user_config([<<"chain">>, Key], Default).
+
+env(Key, Default) ->
+    aeu_env:get_env(aecore, Key, Default).
+
+enabled() ->
+    env(dev_reward_enabled, ?ENABLED).
+beneficiaries() ->
+    env(dev_reward_beneficiaries, ?BENEFICIARIES).
+total_shares() ->
+    ?TOTAL_SHARES.
+allocated_shares() ->
+    case env(dev_reward_allocated_shares, undefined) of
+        undefined ->
+            AllocShares = lists:foldl(fun ({_, X}, A) -> A + X end, 0, beneficiaries()),
+            application:set_env(aecore, dev_reward_allocated_shares, AllocShares),
+            AllocShares;
+        Res ->
+            Res
+    end.
+
+activated(Height) ->
+    case env(dev_reward_activated, undefined) of %% for eunit to avoid mocking
+        undefined ->
+            aec_hard_forks:protocol_effective_at_height(Height) >= ?FORTUNA_PROTOCOL_VSN;
+        Val when is_boolean(Val) ->
+            Val
+    end.
+
+
+unparse_beneficiary({PK, Share}) ->
+    <<(aeser_api_encoder:encode(account_pubkey, PK))/binary, ":", (integer_to_binary(Share))/binary>>.
+
+parse_beneficiary(BeneficiaryShareStr) ->
+    Regex = "^(?'account'ak_[1-9A-HJ-NP-Za-km-z]*):(?'share'[0-9]+)$",
+    case re:split(BeneficiaryShareStr, Regex, [trim, group, {return, binary}]) of
+        [[<<>>, Account, Share]] ->
+            case aeser_api_encoder:safe_decode(account_pubkey, Account) of
+                {ok, PubKey} ->
+                    {PubKey, binary_to_integer(Share)};
+                {error, _} ->
+                    error({invalid_protocol_beneficiary, Account})
+            end;
+        _ ->
+            error({invalid_protocol_beneficiary_share, BeneficiaryShareStr})
+    end.
+
+parse_beneficiaries(BeneficiarySharesStrs) ->
+    try lists:foldl(fun (Str, Acc) -> [parse_beneficiary(Str) | Acc] end,
+                    [], BeneficiarySharesStrs) of
+        Parsed ->
+            {ok, Parsed}
+    catch
+        error:Reason ->
+            {error, Reason}
+    end.
+
+
+split(BeneficiaryReward1, BeneficiaryReward2, NewestNodeHeight) ->
+    case {activated(NewestNodeHeight), enabled()} of
+        {true, true} ->
+            AllocShares = allocated_shares(),
+            TotalShares = total_shares(),
+            AbsContrib1 = BeneficiaryReward1 * AllocShares div TotalShares,
+            AbsContrib2 = BeneficiaryReward2 * AllocShares div TotalShares,
+            DevContrib  = AbsContrib1 + AbsContrib2,
+            {_, DevRewards} =
+                lists:foldl(
+                  fun ({PK, PKShares}, {Remaining, Acc}) ->
+                          Reward0 = DevContrib * PKShares div AllocShares,
+                          Reward1 = min(Reward0, Remaining),
+                          {Remaining - Reward1, [{PK, Reward1} | Acc]}
```

Comment:
Can `Remaining - Reward1` be a negative quantity? `Reward1` is `Remaining` or larger (based on `min` in line above).

Reply:
I think it can't go below 0. If Reward0 is smaller than Remaining than next Remaining is positive, if larger, next Remaining is 0 and stays that way until end of fold.

Reply:
I do not easily see that Reward0 is not larger than Remaining.

If you are confident that `Remaining - Reward1` is never negative quantity, can you please add hardcoded expectation?

And if Reward0 is not larger than Remaining, why having the `min` call in the first place?

Reply:
Thinking in the line of previous comment (^^), the min can be removed, as DIV will cut off the fractional part, so the intermittently accumulated not-assigned-not-locked would end up in Remaining. The implementation when we take this remaining amount (if non-zero), after fold, and add it to last beneficiary should fix this and then all tokens will be distributed.